BennyTordrupVisma/Week2/Ass3/script.js

```diff
@@ -0,0 +1,340 @@
+		var that = this;
+		var listList = this.state.lists.map(function(list, index) {
+			return <List key={index} 
+						list={list} 
+						onArchiveMessage={that.archiveMessage} 
+						onDeleteMessage={that.deleteMessage}
+						onUnarchiveMessage={that.unarchiveMessage}
+						onMoveMessage={that.moveMessage} />
+		});
+		
+		return 	<div>
+					<InputField handleCommit={this.commitMessage} lists={this.state.lists}/>
+					<div>
+						<h3>Lists</h3>
+						<div>
+							{listList}
+						</div>
+					</div>
+				</div>
+	},
+
+	moveMessage: function(message, oldList, newListName) {
+		console.log("Moving message from " + oldList.name + " to " + newListName)
+		var sourceList = this.state.lists.find(l => l.name == oldList.name);
+		if (sourceList == null)
+			return;
+		
+		var destinationList = this.state.lists.find(l => l.name == newListName);
+		if (destinationList == null)
+			return;
+		
+		var msgPos = sourceList.messages.findIndex(m => m.id = message.id);
+		if (msgPos === -1)
+			return;
+
+		var msgText = sourceList.messages[msgPos].text;
+
+		this.deleteMessage(message, oldList);
+		
+		this.commitMessage(msgText, newListName);
+	},
+	
+	deleteMessage: function(message, list) {
+		var listToChange = this.state.lists.find(l => l.name == list.name);
+		if (listToChange == null)
+			return;
+
+		var msgPos = listToChange.messages.findIndex(m => m.id == message.id);
+		if (msgPos === -1)
+			return;
+
+		var tailMessages = listToChange.messages.slice(msgPos + 1);
+		var newMessages = listToChange.messages.slice(0, msgPos).concat(tailMessages);
+		listToChange.messages = newMessages;
+		
+		this.setState({
+			lists: this.state.lists.map(function(list) {
+				return list.name == listToChange.name ? listToChange : list;
+			})
+		});
+	},
+	
+	archiveOrUnarchiveMessageKernel: function(message, list, doArchive) {
+		var listToChange = this.state.lists.find(l => l.name == list.name);
+		if (listToChange == null)
+			return;
+		
+		var msgToArchive = listToChange.messages.find(m => m.id == message.id);
+		if (msgToArchive == null)
+			return;
+		
+		msgToArchive.isArchived = doArchive;
+		
+		this.setState({
+			lists: this.state.lists.map(function(list) {
+					return list.name == listToChange.name ? listToChange : list;
+			})
+		});
+	},
+	
+	archiveMessage: function(message, list) {
+		this.archiveOrUnarchiveMessageKernel(message, list, true);
+	},
+	
+	unarchiveMessage: function(message, list) {
+		this.archiveOrUnarchiveMessageKernel(message, list, false);
+	},
+
+	commitMessage: function(messageText, listName) {
+		var msgToAdd = {
+			id: 0,
+			text: messageText,
+			isArchived: false
+		};
+		
+		var destinationList = this.state.lists.find(l => l.name == listName);	
+		if (destinationList == null) {
+			msgToAdd.id=1;
+			
+			destinationList = {
+				name: listName, 
+				messages: [msgToAdd]
+			};
+			
+			this.setState({
+				lists: this.state && this.state.lists ? this.state.lists.concat([destinationList]) : [destinationList]
```

Comment:
you could implement `getinitialState` and return `{lists: []}` in order to avoid so many checks